rrs/views.py

```diff
@@ -0,0 +1,1005 @@
+
+    @staticmethod
+    def get_reup_by_recipes_and_date(recipes_id, date_id=None):
+        """ Get Recipe Upstream based on Recipes and Recipe Upstream History """
+        stats = []
+
+        if date_id:
+            qry = """SELECT recipesymbol_id, status, no_update_reason, version
+                    FROM rrs_recipeupstream
+                    WHERE history_id = %s
+                    AND recipesymbol_id IN %s;"""
+            cur = connection.cursor()
+            cur.execute(qry, [str(date_id), tuple(recipes_id)])
+            stats = Raw.dictfetchall(cur)
+
+        return stats
+
+    @staticmethod
+    def get_reup_by_last_updated(layerbranch_id, date):
+        """ Get last time the Recipes were upgraded """
+        cur = connection.cursor()
+        cur.execute("""SELECT recipesymbol_id, MAX(commit_date) AS date
+                       FROM rrs_recipeupgrade
+                       INNER JOIN rrs_recipesymbol AS rs
+                       ON rrs_recipeupgrade.recipesymbol_id = rs.id
+                       WHERE commit_date <= %s
+                       AND rs.layerbranch_id = %s
+                       GROUP BY recipesymbol_id;
+                    """, [date, layerbranch_id])
+        return Raw.dictfetchall(cur)
+
+    @staticmethod
+    def get_reup_by_date(date_id):
+        """ Get Recipes not up to date based on Recipe Upstream History """
+        cur = connection.cursor()
+        cur.execute("""SELECT DISTINCT recipesymbol_id
+                        FROM rrs_recipeupstream
+                        WHERE status = 'N'
+                        AND history_id = %s
+                    """, [date_id])
+        return [i[0] for i in cur.fetchall()]
+
+    @staticmethod
+    def get_reupg_by_date(layerbranch_id, date):
+        """ Get info for Recipes for the milestone """
+        cur = connection.cursor()
+        cur.execute("""SELECT rs.id, rs.pn, rs.summary, te.version, rownum FROM (
+                            SELECT recipesymbol_id, version, commit_date, upgrade_type, ROW_NUMBER() OVER(
+                                PARTITION BY recipesymbol_id
+                                ORDER BY commit_date DESC, id DESC
+                            ) AS rownum
+                        FROM rrs_recipeupgrade
+                        WHERE commit_date <= %s
+                        AND upgrade_type not in ('M', 'N')) AS te
+                        INNER JOIN rrs_recipesymbol AS rs
+                        ON te.recipesymbol_id = rs.id
+                        WHERE rownum = 1
+                        AND te.upgrade_type <> 'R'
+                        AND rs.layerbranch_id = %s
+                        ORDER BY rs.pn;
+                        """, [date, layerbranch_id])
+        return Raw.dictfetchall(cur)
+
+    @staticmethod
+    def get_reupg_by_dates_and_recipes(start_date, end_date, recipes_id):
+        """  Get Recipe Upgrade for the milestone based on Recipes """
+        cur = connection.cursor()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *cur*. The resource is allocated by call *connection.cursor*. Execution paths that do not contain closure statements were detected. To prevent this resource leak, close *cur* in a try-finally block or declare it using a `with` statement.

**More info**
[View details about the `with` statement in the Python developer&#39;s guide](https://www.python.org/dev/peps/pep-0343/) (external link).